### DISCUSSION TITLE
Backporting of Mapbox 2.x bug fixes

### DISCUSSION THREAD

**Discussion by dannybloe** [maplibre/maplibre-gl-js#701] [General]
Hi folks, 

In mapbox there was a bug fix merged for this bug https://github.com/mapbox/mapbox-gl-js/issues/10491. Is it possible to merge that into maplibre as well? We just found out that we have to revert to an older version of mapbox because the fix (pr [Always retain raster tile cache control data #10494](https://github.com/mapbox/mapbox-gl-js/pull/10494/)) was only merged in 2.x branch so it never ended up in maplibre. 

That was a bit of a bummer. We could get it working by doing `window.OffscreenCanvas = null;` but then other things break.

Cheers, 
Danny

**Comment by HarelM** [comment]
We can't copy code from mapbox 2.x due to licensing issues.
If you know who sent the PR and if they are willing to do it here that's another story...

**Comment by jaapster** [comment]
In what way would that be different from copying from Mapbox 2.x? The resulting code would be similar.

**Comment by dannybloe** [comment]
This is a difficult one. That would almost imply this bug cannot be fixed. Btw, the fix was made by @andycalder who is not a mapbox employee so it seems.

**Comment by HarelM** [comment]
>In what way would that be different from copying from Mapbox 2.x? The resulting code would be similar.

The fact that the resulting code would be similar is not a problem if the person who is writing the code never looked at the code in 2.x that solves this issue.
If you fix the code in the same way but have never looked at mapbox's code then it does not mean you copied.
If any of you know how to fix this issue without looking at the mapbox version 2.x code and can submit a PR it's valid too.

**Comment by wipfli** [comment]
@nyurik, @klokan; would be great if you could share your view on this.

**Comment by dannybloe** [comment]
So if some function should do `return true;` in 2.x in order to fix a bug then anybody who ever saw the code of 2.x could never fix that same bug in that same equivalent function in maplibre by also returning `true` there?

**Comment by wipfli** [comment]
@dannybloe thanks for asking this question. I think it is a super important one, and we should discuss about it in the discussion section.

**Comment by nyurik** [comment]
I second many sentiments by @HarelM and @wipfli here.  My IANAL perspective:
* we MUST NOT copy any code from 2.x or any other non-compatible licensed code.
* the author of any code can contribute it under multiple licenses, so if @andycalder can contribute the same patch under BSD license, we will be very happy if they do. But no pressure! :)
* we can re-implement any functionality we need, as long as the person writing the code is not looking at the other code at the same time. This is not the same as "I have seen the non-BSD code before, therefore I can never write anything similar because i am tainted". This is more of a "I am making a reasonable effort to implement similar functionality using overall problem description and specifications as a guide, but I am not infringing on someone else's work by copying it".
* From my recollection, trivial code one-liners are not considered infringements (but theoretically a comment might be). I think there has to be substantial copying for anyone to really object, but obviously this is a trickier matter of how many lines would mean infringement.

In short, don't copy. Write your own. It is normal that similar goals might be achieved by similar implementations (there are only so many ways one can write a `min(a, b)` function).

**Reply by dannybloe** [reply]
Yes, this makes sense indeed.

But we still have this bug that prevents us from using maplibre which is a bummer (but that's not an issue for this ticket).

**Comment by Pictor13** [comment]
If I understood correctly the physical good here is the copyrighted code (digital implementation of a concept), while the idea & understanding & functionality (abstract implementation of a concept) has no licensing issues.

So I'm just wondering...
what if _DevFoo_ looks at the _Mapbox_ code, understands the bug & the fix, to then explain (in words) what the fix is about to _DevBar_, without providing the precise code?
It's highly unlikely that a fix from _DevBar_ would be coded in the same exact way as _Mapbox_, especially on a multiple-lines fix.
This workflow would still allow to look at the Mapbox code as "inspiration" for a fix (so to at least not duplicate the debugging effort).

Of course this requires 2 entities/devs, rather than only one.

**Reply by wipfli** [reply]
One of the biggest threats to the MapLibre Organization is copying code from Mapbox. Let us keep this in mind in this discussion.

MapLibre is and independent effort, we go our way, and our community can make beautiful things happen.

**Comment by randName** [comment]
started using maplibre recently and just came across this discussion; in case anyone is like me and wanted to fix the original bug in question, it was already fixed by #1106
